<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.2"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran.js: Single cell RNA-seq analysis in Javascript</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">scran.js
   </div>
   <div id="projectbrief">Compiling scran&#39;s C++ library to WebAssembly</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.2 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">Single cell RNA-seq analysis in Javascript </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p ><a class="anchor" id="md__github_workspace_README"></a> </p><h1>Overview</h1>
<p >This repository contains <b>scran.js</b>, a Javascript library for single-cell RNA-seq (scRNA-seq) analysis in the browser. The various calculations are performed directly by the client, allowing us to take advantage of the ubiquity of the browser as a standalone analysis platform. Users can then directly analyze their data without needing to manage any dependencies or pay for access to a backend. <b>scran.js</b> is heavily inspired by the <a href="https://bioconductor.org/packages/scran"><b>scran</b> R package</a> and contains most of its related methods. Indeed, much of the implementation in this repository is taken directly from <b>scran</b> and its related R packages.</p>
<h1>Key scRNA-seq analysis steps</h1>
<p >Currently, the library and web app supports the key steps in a typical scRNA-seq analysis:</p>
<ul>
<li>Quality control (QC) to remove low-quality cells. This is done based on detection of outliers on QC metrics like the number of detected genes.</li>
<li>Normalization and log-transformation, to remove biases and mitigate the mean-variance trend. We use scaling normalization with size factors defined from the library size for each cell.</li>
<li>Feature selection to identify highly variable genes. This is based on residuals from a trend fitted to the means and variances of the log-normalized data for each gene.</li>
<li>Principal components analysis (PCA) on the highly variable genes, to compress and denoise the data. We use an approximate method to quickly obtain the top few PCs.</li>
<li>Clustering using multi-level community detection (a.k.a., "Louvain clustering"). This is performed on the top PCs.</li>
<li>Dimensionality reduction with t-stochastic neighbor embedding (t-SNE), again using the top PCs.</li>
<li>Marker detection using a variety of effect sizes such as Cohen's d and the area under the curve (AUC).</li>
</ul>
<p >Coming soon:</p>
<ul>
<li>Clustering using k-means.</li>
<li>Dimensionality reduction by uniform map and approximate projection (UMAP).</li>
<li>Batch correction via the mutual nearest neighbors method.</li>
</ul>
<p >The theory behind these methods is described in more detail in the <a href="https://bioconductor.org/books/release/OSCA/"><b>Orchestrating Single Cell Analysis with Bioconductor</b></a> book.</p>
<h1>Efficient analysis with WebAssembly</h1>
<p >We use WebAssembly (Wasm) to enable efficient client-side execution of common steps in a scRNA-seq analysis. Code to perform each step is written in C++ and compiled to Wasm using the <a href="https://emscripten.org/">Emscripten toolchain</a>. Some of the relevant C++ libraries are listed below:</p>
<ul>
<li><a href="https://github.com/LTLA/libscran">libscran</a> provides C++ implementations of key functions in <b>scran</b> and its fellow packages <b>scater</b> and <b>scuttle</b>. This includes quality control, normalization, feature selection, PCA, clustering and dimensionality reduction.</li>
<li><a href="https://github.com/LTLA/tatami">tatami</a> provides an abstract interface to different matrix classes, focusing on row and column extraction.</li>
<li><a href="https://github.com/LTLA/knncolle">knncolle</a> wraps a number of nearest neighbor detection methods in a consistent interface.</li>
<li><a href="https://github.com/LTLA/CppIrlba">CppIrlba</a> contains a C++ port of the IRLBA algorithm for approximate PCA.</li>
<li><a href="https://github.com/LTLA/CppKmeans">CppKmeans</a> contains C++ ports of the Hartigan-Wong and Lloyd algorithms for k-means clustering.</li>
<li><a href="https://github.com/LTLA/qdtsne">qdtsne</a> contains a refactored C++ implementation of the Barnes-Hut t-SNE dimensionality reduction algorithm.</li>
<li><a href="https://github.com/LTLA/umappp">umappp</a> contains a refactored C++ implementation of the UMAP dimensionality reduction algorithm.</li>
</ul>
<p >For each step, we use Emscripten to compile the associated C++ functions into Wasm and generate Javascript-visible bindings. We can then load the Wasm binary into a web application and call the desired functions on user-supplied data.</p>
<h1>Building the Wasm binary</h1>
<p >This directory contains the files required to create the <b>scran.js</b> Wasm binary. We use CMake to manage the compilation process as well as the dependencies, namely the <a href="https://github.com/LTLA/libscran"><b>scran</b> C++ library</a>. Compilation of the Wasm binary is done using Emscripten:</p>
<div class="fragment"><div class="line">emcmake cmake -S . -B build</div>
<div class="line">(cd build &amp;&amp; emmake make)</div>
</div><!-- fragment --><p >As a npm package</p>
<div class="fragment"><div class="line">emcmake cmake -S . -B build</div>
<div class="line">(cd build &amp;&amp; emmake make)</div>
</div><!-- fragment --><p >This will build the <code>.js</code> and <code>.wasm</code> file within the <code>build/</code> subdirectory.</p>
<h1>Developer notes</h1>
<p >Make sure emscripten and cmake are installed on your machine</p>
<p >To build the node version o the library</p>
<div class="fragment"><div class="line">bash build.node.sh</div>
</div><!-- fragment --><p >To build normal version of the library</p>
<div class="fragment"><div class="line">bash build.sh</div>
</div><!-- fragment --><h2>Tests</h2>
<div class="fragment"><div class="line"># install dev dependencies</div>
<div class="line">npm install --include=dev</div>
<div class="line">npm run test</div>
</div><!-- fragment --><p ><em><b>Note: This dev notes is tested with node v17.2.0</b></em> </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.2
</small></address>
</body>
</html>
