<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran.js: Single cell RNA-seq analysis in Javascript</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran.js
   </div>
   <div id="projectbrief">Compiling scran&#39;s C++ library to WebAssembly</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">Single cell RNA-seq analysis in Javascript </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p ><a class="anchor" id="md__github_workspace_README"></a> </p><h1>Overview</h1>
<p >This repository contains <b>scran.js</b>, a Javascript library for single-cell RNA-seq (scRNA-seq) analysis in the browser. The various calculations are performed directly by the client, allowing us to take advantage of the ubiquity of the browser as a standalone analysis platform. Users can then directly analyze their data without needing to manage any dependencies or pay for access to a backend. <b>scran.js</b> is heavily inspired by the <a href="https://bioconductor.org/packages/scran"><b>scran</b> R package</a> and contains most of its related methods. Indeed, much of the implementation in this repository is taken directly from <b>scran</b> and its related R packages.</p>
<h1>Analysis overview</h1>
<p >Currently, the library and web app supports the key steps in a typical scRNA-seq analysis:</p>
<ul>
<li>Quality control (QC) to remove low-quality cells. This is done based on detection of outliers on QC metrics like the number of detected genes.</li>
<li>Normalization and log-transformation, to remove biases and mitigate the mean-variance trend. We use scaling normalization with size factors defined from the library size for each cell.</li>
<li>Feature selection to identify highly variable genes. This is based on residuals from a trend fitted to the means and variances of the log-normalized data for each gene.</li>
<li>Principal components analysis (PCA) on the highly variable genes, to compress and denoise the data. We use an approximate method to quickly obtain the top few PCs.</li>
<li>Clustering using multi-level community detection (a.k.a., "Louvain clustering") on a shared nearest neighbor graph. This is performed on the top PCs.</li>
<li>Dimensionality reduction with t-stochastic neighbor embedding (t-SNE), again using the top PCs.</li>
<li>Marker detection using a variety of effect sizes such as Cohen's d and the area under the curve (AUC).</li>
</ul>
<p >Coming soon:</p>
<ul>
<li>Clustering using k-means.</li>
<li>Batch correction via the mutual nearest neighbors method.</li>
<li>Cell type annotation with <b>SingleR</b>.</li>
</ul>
<p >The theory behind these methods is described in more detail in the <a href="https://bioconductor.org/books/release/OSCA/"><b>Orchestrating Single Cell Analysis with Bioconductor</b></a> book.</p>
<h1>Quick start</h1>
<p ><b>scran.js</b> is available as an <a href="https://www.npmjs.com/package/scran.js">npm package</a>, so installation can be performed via the usual procedure:</p>
<div class="fragment"><div class="line">npm install scran.js</div>
</div><!-- fragment --><p >Then you can import and initialize the library using the standard ES6 notation. Before any actual analysis steps are performed, the <code>initialize()</code> function must be run and its promise resolved; we suggest using a top-level <code>await</code> for convenience.</p>
<div class="fragment"><div class="line">import * as scran from &quot;scran.js&quot;;</div>
<div class="line">await scran.initialize({ numberOfThreads: 4 }); // TODO: for node, set localFile: true</div>
</div><!-- fragment --><p >After that, you can run the remaining steps synchronously. Here is an example using the Node.js API with one of our example <a href="https://github.com/jkanche/random-test-files">Matrix Market files</a>:</p>
<div class="fragment"><div class="line">// Reading the file in.</div>
<div class="line">import * as fs from &quot;fs&quot;;</div>
<div class="line">let buffer = fs.readFileSync(&quot;matrix.mtx.gz&quot;);</div>
<div class="line">let mat = scran.initializeSparseMatrixFromMatrixMarketBuffer(buffer);</div>
<div class="line"> </div>
<div class="line">// Performing the QC.</div>
<div class="line">let qc_metrics = scran.computePerCellQCMetrics(mat, [ /* specify mito subset here */ ]);</div>
<div class="line">let qc_thresholds = scran.computePerCellQCThresholds(qc_metrics);</div>
<div class="line">let filtered = scran.filterCells(mat, qc_thresholds.discardOverall());</div>
<div class="line"> </div>
<div class="line">// Log-normalizing.</div>
<div class="line">let normalized = scran.logNormCounts(filtered);</div>
<div class="line"> </div>
<div class="line">// Modelling per-gene variance and selecting top HVGs (TODO: make this easier).</div>
<div class="line">let varmodel = scran.modelGeneVar(normalized);</div>
<div class="line"> </div>
<div class="line">let resid_copy = varmodel.residuals().slice();</div>
<div class="line">resid_copy.sort();</div>
<div class="line">let threshold = resid_copy[resid_copy.length - 4000]; // top 4000 HVGs.</div>
<div class="line">let features = new Uint8Array(resid_copy.length);</div>
<div class="line">varmodel.residuals().forEach((x, i) =&gt; {</div>
<div class="line">    features[i] = x &gt;= threshold;</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line">// Performing the PCA.</div>
<div class="line">let pcs = scran.runPCA(normalized, { features: features });</div>
<div class="line"> </div>
<div class="line">// Building the neighbor search index on the PCs.</div>
<div class="line">let index = scran.buildNeighborSearchIndex(pcs);</div>
<div class="line"> </div>
<div class="line">// Performing the clustering. (TODO: allow direct use of index in buildSNNGraph).</div>
<div class="line">let cluster_nns = scran.findNearestNeighbors(index, 10);</div>
<div class="line">let cluster_graph = scran.buildSNNGraph(cluster_nns);</div>
<div class="line">let clustering  = scran.clusterSNNGraph(cluster_graph);</div>
<div class="line"> </div>
<div class="line">// Performing the t-SNE and UMAP.</div>
<div class="line">let tsne_status = scran.initializeTSNE(index);</div>
<div class="line">let tsne_res = scran.runTSNE(tsne_status);</div>
<div class="line"> </div>
<div class="line">let umap_status = scran.initializeUMAP(index);</div>
<div class="line">let umap_res = scran.runUMAP(umap_status);</div>
</div><!-- fragment --><p >On Node.js, it may also be necessary to terminate the workers so that the process can shut down properly; this is achieved by calling <code>scran.terminate()</code> once all operations are finished.</p>
<h1>Developer notes</h1>
<h2>Introducing WebAssembly</h2>
<p >We use WebAssembly (Wasm) to enable efficient client-side execution of common steps in a scRNA-seq analysis. Code to perform each step is written in C++ and compiled to Wasm using the <a href="https://emscripten.org/">Emscripten toolchain</a>. Some of the relevant C++ libraries are listed below:</p>
<ul>
<li><a href="https://github.com/LTLA/libscran">libscran</a> provides C++ implementations of key functions in <b>scran</b> and its fellow packages <b>scater</b> and <b>scuttle</b>. This includes quality control, normalization, feature selection, PCA, clustering and dimensionality reduction.</li>
<li><a href="https://github.com/LTLA/tatami">tatami</a> provides an abstract interface to different matrix classes, focusing on row and column extraction.</li>
<li><a href="https://github.com/LTLA/knncolle">knncolle</a> wraps a number of nearest neighbor detection methods in a consistent interface.</li>
<li><a href="https://github.com/LTLA/CppIrlba">CppIrlba</a> contains a C++ port of the IRLBA algorithm for approximate PCA.</li>
<li><a href="https://github.com/LTLA/CppKmeans">CppKmeans</a> contains C++ ports of the Hartigan-Wong and Lloyd algorithms for k-means clustering.</li>
<li><a href="https://github.com/LTLA/qdtsne">qdtsne</a> contains a refactored C++ implementation of the Barnes-Hut t-SNE dimensionality reduction algorithm.</li>
<li><a href="https://github.com/LTLA/umappp">umappp</a> contains a refactored C++ implementation of the UMAP dimensionality reduction algorithm.</li>
</ul>
<p >For each step, we use Emscripten to compile the associated C++ functions into Wasm and generate Javascript-visible bindings. We can then load the Wasm binary into a web application and call the desired functions on user-supplied data.</p>
<h2>Building the Wasm binary</h2>
<p >Make sure <a href="https://emscripten.org/docs/getting_started/downloads.html">Emscripten</a> and <a href="https://cmake.org/download/">CMake</a> are installed on your machine. Running the <code>build.sh</code> script will then generate ES6 and Node.js-compatible builds. To build the Node.js version:</p>
<div class="fragment"><div class="line">bash build.sh main</div>
</div><!-- fragment --><p >To build the ES6 version:</p>
<div class="fragment"><div class="line">bash build.sh module</div>
</div><!-- fragment --><p >This will create the <code>main</code> and <code>module</code> directories respectively, containing the Wasm file in the <code>wasm</code> subdirectory as well as copying all the relevant Javascript bindings.</p>
<h2>Tests</h2>
<p >Run the test suite by calling:</p>
<div class="fragment"><div class="line"># install dev dependencies</div>
<div class="line">npm install --include=dev</div>
<div class="line">npm run test</div>
</div><!-- fragment --><p >If you are using earlier versions of Node, instead run tests using:</p>
<div class="fragment"><div class="line">node --experimental-vm-modules --experimental-wasm-threads node_modules/jest/bin/jest.js</div>
</div><!-- fragment --><h1>Links</h1>
<p >Check out our <a href="https://github.com/jkanche/kana">Kana Application</a> to see how <b>scran.js</b> package is used in an interactive scRNA-seq analysis application. </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
